### PR TITLE
Null string is encoded as "null" in incremental index

### PR DESCRIPTION
**Pull request by navis**
I've found null values for string type is encoded as "null", which seemed not intended behavior.

### CODE REVIEW

**Comment by gianm** [comment]
Is it possible to include a test that would have caught this?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 1]
```diff
@@ -98,17 +97,19 @@
 
   private static final Function<Object, String> STRING_TRANSFORMER = new Function<Object, String>()
   {
+    @Nullable
     @Override
-    public String apply(final Object o)
+    public String apply(@Nullable final Object o)
     {
-      return String.valueOf(o);
+      return o == null ? null : String.valueOf(o);
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: can you explain in your PR description why this line is not enough to fix the problem. What is the purpose of the remaining changes?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 2]
```diff
@@ -965,6 +966,7 @@ public DateTime getMaxIngestedEventTime()
     private final int index;
     private final String name;
     private final DimDim values;
+    private final Function transformer;
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: this change seems unrelated to the description in the PR

**Comment by jon-wei** [comment]
Can we remove the changes not related to the String null handling?

The code being refactored is being redone in https://github.com/druid-io/druid/pull/2760, e.g. those value transformers in IncrementalIndex are removed.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: Yes, I've found this doing something else. I'll remove unrelated codes.

**Comment by navis** [comment]
@jon-wei sure

**Comment by drcrallen** [comment]
Is this a bug that affects 0.9.0, or just 0.9.1?

**Comment by navis** [comment]
@drcrallen it's from #2263, which is not included in any released versions.

**Comment by jon-wei** [comment]
@drcrallen it'll just be for 0.9.1, the code being changed was from https://github.com/druid-io/druid/pull/2263, which isn't in 0.9.0

**Comment by jon-wei** [comment]
:+1:

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 3]
```diff
@@ -101,7 +101,7 @@
     @Override
     public String apply(final Object o)
     {
-      return String.valueOf(o);
+      return o == null ? null : String.valueOf(o);
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: Should this be converting `""` to `null` or is something else doing that?

Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: I think it's fine, null/empty dim values will end up as "" via a nullToEmpty() call when added to the dictionary:

https://github.com/navis/druid/blob/invalid-encode-nullstring/processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java#L1125

Comment (gianm) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: ah ok, this looks good then. thanks

**Comment by gianm** [comment]
:+1:

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 4]
```diff
@@ -114,8 +114,9 @@ public Long apply(final Object o)
         return null;
       }
       if (o instanceof String) {
+        String s = (String) o;
         try {
-          return Long.valueOf((String) o);
+          return s.isEmpty() ? null : Long.valueOf(s);
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: does this change any current behavior that might map empty strings to 0 for numeric values?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 5]
```diff
@@ -137,8 +138,9 @@ public Float apply(final Object o)
         return null;
       }
       if (o instanceof String) {
+        String s = (String) o;
         try {
-          return Float.valueOf((String) o);
+          return s.isEmpty() ? null : Float.valueOf(s);
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: let's make sure we have a test that would have failed without this change

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 6]
```diff
@@ -355,6 +357,9 @@ public int lookupId(String name)
   private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
   private final List<DimDim> dimValues;
 
+  // looks need a configuration
+  private final Ordering<Comparable> ordering = Ordering.natural().nullsFirst();
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: do we have a test for ordering somewhere?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 4 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: I don't think it could, since with the old code, `Long.valueOf("")` would have thrown an exception. The new code would return null.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 6 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: @jon-wei I'm not sure I understand what the comment means, can we have a test for this?

Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: @xvrl I think this should be directed at @navis

Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: @jon-wei my bad
